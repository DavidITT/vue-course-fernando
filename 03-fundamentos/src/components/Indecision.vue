<template>
  <h1>Indesicion</h1>
  <img v-if="image" :src="image" alt="image">
  <div class="bg-dark"></div>
  <div class="indecision-container">
    <input type="text" placeholder="Realiza una pregunta" v-model="question">
    <p><b>Nota: Recuerda poner un signo de admiraci√≥n al final (?).</b></p>
    <div v-if="isValid">
      <h2>{{ question }}</h2>
      <h1>{{ answer }}</h1>
    </div>
  </div>
</template>

<script>
export default {
  name: "Indecision",

  data() {
    return {
      question: null,
      answer: null,
      image: null,
      isValid: false,
    }
  },
  methods: {
    async getAnswer() {
      try {
        this.answer = 'Pensando...'
        const {answer, image} = await fetch('https://yesno.wtf/api').then(res => res.json())
        this.answer = answer === 'yes' ? 'Si' : 'No'
        this.image = image
      } catch (error) {
        this.answer = 'No se pudo cargar'
        this.image = null
      }
    }
  },
  watch: {
    question(value, oldValue) {
      this.isValid = false
      console.log({value})
      if (!value.includes('?')) return
      console.log({value})
      //TODO: Realizar peticion http
      this.isValid = true
      this.getAnswer()
    }
  }
}
</script>

<style scoped>

img, .bg-dark {
  height: 100vh;
  left: 0px;
  max-height: 100%;
  max-width: 100%;
  position: fixed;
  top: 0px;
  width: 100vw;
}

.bg-dark {
  background-color: rgba(0, 0, 0, 0.4);
}

.indecision-container {
  position: relative;
  z-index: 99;
}

input {
  width: 250px;
  padding: 10px 15px;
  border-radius: 5px;
  border: none;
  margin-bottom: 10px;
}

input:focus {
  outline: none;
}

p {
  color: white;
  font-size: 20px;
  margin-top: 0px;
}

h1, h2 {
  color: white;
}

h2 {
  margin-top: 150px;
}

</style>