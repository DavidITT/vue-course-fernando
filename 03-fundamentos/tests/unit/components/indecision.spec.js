import {shallowMount} from '@vue/test-utils'
import Indecision from "@/components/Indecision"


describe('Indecision component', () => {

    let wrapper
    let clgSpy

    global.fetch = jest.fn(() => Promise.resolve({
        json: () => Promise.resolve({
            'answer': 'yes',
            'forced': false,
            'image': 'https://yesno.wtf/assets/yes/7-653c8ee5d3a6bbafd759142c9c18d76c.gif'
        })
    }))

    beforeEach(() => {
        wrapper = shallowMount(Indecision)
        clgSpy = jest.spyOn(console, 'log')

        jest.clearAllMocks()
    })

    test('debe de hacer match con el snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('Escribir en el input no debe de disparar nada (console.log)', async () => {

        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')
        const input = wrapper.find('input')
        await input.setValue('Hola mundo')

        expect(clgSpy).toHaveBeenCalledTimes(1)
        expect(getAnswerSpy).not.toHaveBeenCalled()

        //console.log(wrapper.vm)

    })

    test('Al escrbir el simbolo (?) se debe deisparar el getAnswer', async () => {

        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')
        const input = wrapper.find('input')
        await input.setValue('Hola mundo?')

        expect(clgSpy).toHaveBeenCalledTimes(2)
        expect(getAnswerSpy).toHaveBeenCalled()
    })

    test('Pruebas en getAnswer()', async () => {
        await wrapper.vm.getAnswer()
        const img = wrapper.find('img')

        expect( img.exists() ).toBeTruthy()
        expect( wrapper.vm.image ).toBe('https://yesno.wtf/assets/yes/7-653c8ee5d3a6bbafd759142c9c18d76c.gif')
        expect( wrapper.vm.answer ).toBe('Si')
        //console.log(wrapper.vm.image)
        //console.log(wrapper.vm.answer)
    })

    test('Pruebas en getAnswer() - Fallo en el API', async () => {

        //TODO: Fallo en el API
        fetch.mockImplementationOnce( () => Promise.reject('API is down'))

        await wrapper.vm.getAnswer()

        const img = wrapper.find('img')

        expect( img.exists() ).toBeFalsy()
        expect( wrapper.vm.answer ).toBe('No se pudo cargar')

    })


})