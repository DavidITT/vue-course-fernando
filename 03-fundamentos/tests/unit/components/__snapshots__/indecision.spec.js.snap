// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision component debe de hacer match con el snapshot 1`] = `
<h1>Indesicion</h1>
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Realiza una pregunta">
  <p><b>Nota: Recuerda poner un signo de admiraci√≥n al final (?).</b></p>
  <!--v-if-->
</div>
`;
